File: bindings/aitp-py/tests/test_revocation.py
"""Revocation-list signing — Python SDK.

Round-trip: build a list with two entries, parse it back, verify the
issuer signature, and confirm the entries land in expected positions.
"""

import json
import uuid

import aitp


def test_sign_revocation_list_round_trips():
    issuer = aitp.AitpAgent.generate()
    issuer.build_manifest(
        display_name="issuer",
        handshake_endpoint="http://localhost:8100/aitp/handshake/",
        offered_caps=["demo.echo"],
    )

    jti_a = str(uuid.uuid4())
    jti_b = str(uuid.uuid4())
    envelope_json = issuer.sign_revocation_list(
        [
            {"jti": jti_a, "reason": "compromised"},
            {"jti": jti_b, "revoked_at": 1_700_000_000},
        ],
        expires_in_secs=600,
    )

    env = json.loads(envelope_json)
    body = env["revocation_list"]
    assert body["issuer"] == issuer.aid
    assert body["version"] == "aitp/0.2"
    assert len(body["entries"]) == 2

    jti_set = {e["jti"] for e in body["entries"]}
    assert jti_set == {jti_a, jti_b}

    # The custom revoked_at on entry B must survive the round-trip.
    by_jti = {e["jti"]: e for e in body["entries"]}
    assert by_jti[jti_b]["revoked_at"] == 1_700_000_000
    assert by_jti[jti_a]["reason"] == "compromised"

    # Expiry is now + 600.
    assert body["expires_at"] - body["published_at"] == 600

    # Envelope carries a signature string.
    assert isinstance(env["signature"], str) and env["signature"]


def test_sign_revocation_list_rejects_bad_uuid():
    issuer = aitp.AitpAgent.generate()
    issuer.build_manifest(
        display_name="issuer",
        handshake_endpoint="http://localhost:8100/aitp/handshake/",
        offered_caps=["demo.echo"],
    )
    try:
        issuer.sign_revocation_list([{"jti": "not-a-uuid"}])
    except ValueError:
        return
    raise AssertionError("sign_revocation_list accepted a bad UUID")


def test_sign_revocation_list_requires_jti():
    issuer = aitp.AitpAgent.generate()
    issuer.build_manifest(
        display_name="issuer",
        handshake_endpoint="http://localhost:8100/aitp/handshake/",
        offered_caps=["demo.echo"],
    )
    try:
        issuer.sign_revocation_list([{"reason": "missing jti"}])
    except ValueError:
        return
    raise AssertionError("sign_revocation_list accepted an entry missing jti")


# ── Verification ─────────────────────────────────────────────────────────
#
# The verify half was missing until 0.6.0: both the Python and Node bindings
# exposed `sign_revocation_list` and neither exposed `verify_revocation_list`,
# even though `verify_revocation_snapshot` is a Tier C conformance operation
# the Rust adapter implements. Three downstream repos then hand-rolled,
# skipped, or faked verification, and the 0.5.0 signing-input change crossed a
# whole release family with one accidental interlock in its way.


def _kat_issuer_aid() -> str:
    """The KAT keypair's AID, read from `keypairs.json` — NOT from the
    snapshot being verified.

    Taking it from the envelope would make the issuer-binding half of
    verification tautological: a snapshot whose `issuer` had been swapped
    (with a matching signature) would still pass.
    """
    import pathlib

    root = pathlib.Path(__file__).resolve().parents[3]
    keypairs = json.loads(
        (root / "tests/schemas/known-answer/keypairs.json").read_text()
    )

    def find(node):
        if isinstance(node, dict):
            if node.get("id") == "kat-keypair-001" and "aid" in node:
                return node["aid"]
            for value in node.values():
                found = find(value)
                if found:
                    return found
        elif isinstance(node, list):
            for value in node:
                found = find(value)
                if found:
                    return found
        return None

    aid = find(keypairs)
    assert aid, "kat-keypair-001 not found in keypairs.json"
    return aid


def _committed_snapshot() -> str:
    import pathlib

    root = pathlib.Path(__file__).resolve().parents[3]
    doc = json.loads(
        (
            root
            / "tests/schemas/known-answer/signed-examples/revocation"
            / "kat-keypair-001-snapshot.json"
        ).read_text()
    )
    # `_kat_input` is a minting companion, never part of the wire object.
    doc.pop("_kat_input", None)
    return json.dumps(doc)


def test_verify_revocation_list_accepts_a_freshly_signed_snapshot():
    issuer = aitp.AitpAgent.generate()
    envelope = issuer.sign_revocation_list([{"jti": str(uuid.uuid4())}], 600)
    assert aitp.verify_revocation_list(envelope, issuer.aid) is None


def test_verify_revocation_list_verifies_the_committed_spec_vector_as_committed():
    """Cross-implementation check — the one that would have caught 0.5.0.

    The vector is verified **as committed**, without re-minting. A suite where
    the same code signs and verifies passes under any self-consistent
    convention, including a wrong one; that self-consistency is precisely what
    let an aitp-rs issuer and an aitp-verifier-py consumer disagree on the wire
    while both suites stayed green.
    """
    assert (
        aitp.verify_revocation_list(
            _committed_snapshot(), _kat_issuer_aid(), 1_711_900_100
        )
        is None
    )


def _code_for(envelope_json: str, issuer_aid: str, now=None) -> str:
    try:
        aitp.verify_revocation_list(envelope_json, issuer_aid, now)
    except aitp.RevocationVerificationError as exc:
        return exc.code
    raise AssertionError("expected verification to fail, but it succeeded")


def test_every_failure_cause_is_recoverable_without_parsing_a_message():
    """`.code` is the contract; the message wording is not.

    A caller forced to string-match an exception message is pinning program
    output as an expected value — the bug class this whole effort exists to
    remove. Before this binding existed, that was the only option available.
    """
    issuer = aitp.AitpAgent.generate()
    other = aitp.AitpAgent.generate()
    envelope = issuer.sign_revocation_list([{"jti": str(uuid.uuid4())}], 600)

    assert _code_for(envelope, other.aid) == "issuer_mismatch"

    tampered = json.loads(envelope)
    tampered["revocation_list"]["entries"] = []
    assert _code_for(json.dumps(tampered), issuer.aid) == "signature_invalid"

    bad_version = json.loads(envelope)
    bad_version["revocation_list"]["version"] = "aitp/9.9"
    assert _code_for(json.dumps(bad_version), issuer.aid) == "version_unknown"

    assert _code_for(envelope, issuer.aid, 99_999_999_999) == "expired"
    assert _code_for("not json at all", issuer.aid) == "malformed"


def test_issuer_mismatch_is_distinct_from_malformed():
    """These were the same cause until 0.6.0 (both `CnfMalformed`).

    An attacker serving their own correctly-signed list and a corrupt fetch
    are different events; a caller that cannot separate them cannot alert on
    the first without drowning in the second.
    """
    issuer = aitp.AitpAgent.generate()
    other = aitp.AitpAgent.generate()
    envelope = issuer.sign_revocation_list([{"jti": str(uuid.uuid4())}], 600)

    # Internally valid — it verifies against its own issuer.
    assert aitp.verify_revocation_list(envelope, issuer.aid) is None
    assert _code_for(envelope, other.aid) == "issuer_mismatch"
    assert _code_for("{}", issuer.aid) == "malformed"


def test_a_bad_expected_issuer_is_the_callers_error_not_a_verification_cause():
    issuer = aitp.AitpAgent.generate()
    envelope = issuer.sign_revocation_list([{"jti": str(uuid.uuid4())}], 600)
    try:
        aitp.verify_revocation_list(envelope, "not-an-aid")
    except aitp.RevocationVerificationError:  # pragma: no cover
        raise AssertionError(
            "a malformed expected_issuer_aid is the caller's mistake; "
            "reporting it as a snapshot verification cause would blame the "
            "wrong party"
        )
    except ValueError:
        pass
    else:  # pragma: no cover
        raise AssertionError("expected a ValueError")


def test_revocation_signing_bytes_are_the_inner_body_not_the_wrapper():
    """The 0.5.0 convention, exposed so callers stop reconstructing it.

    Reconstructing the shape at the call site is exactly how signer, verifier
    and conformance fixture drifted apart before 0.5.0.
    """
    issuer = aitp.AitpAgent.generate()
    envelope = issuer.sign_revocation_list([{"jti": str(uuid.uuid4())}], 600)
    body = json.loads(envelope)["revocation_list"]

    signing_bytes = aitp.revocation_signing_bytes(envelope)

    assert signing_bytes.startswith(b"{"), signing_bytes[:20]
    assert b'"revocation_list"' not in signing_bytes[:32], (
        "the signing input starts with the transport wrapper — this is the "
        "pre-0.5.0 shape"
    )
    # Derived from the envelope, never pasted from output.
    assert signing_bytes == json.dumps(
        body, separators=(",", ":"), sort_keys=True
    ).encode()


def test_a_wrapped_form_signature_is_rejected():
    """The 0.5.0 break, asserted from the verifier's side.

    Up to 0.4.x the signing input was `JCS({"revocation_list": body})` — the
    transport wrapper. From 0.5.0 it is the inner body alone, and there is no
    dual-accept in either direction. This mints a snapshot the *old* way, with
    a local key rather than the SDK, and requires that it fail: an SDK that
    accepted it would silently re-open the wire incompatibility.

    The signer here is deliberately independent of the code under test — a
    test where the SDK both signs and verifies passes under any
    self-consistent convention, including a wrong one.
    """
    import base64
    import hashlib
    import time

    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

    private = Ed25519PrivateKey.generate()
    raw_pub = private.public_key().public_bytes_raw()
    aid = "aid:pubkey:" + base64.urlsafe_b64encode(raw_pub).decode().rstrip("=")

    now = int(time.time())
    body = {
        "version": "aitp/0.2",
        "issuer": aid,
        "published_at": now,
        "expires_at": now + 600,
        "entries": [{"jti": str(uuid.uuid4()), "revoked_at": now}],
    }
    wrapped_input = json.dumps(
        {"revocation_list": body}, separators=(",", ":"), sort_keys=True
    ).encode()
    signature = private.sign(hashlib.sha256(wrapped_input).digest())

    legacy = json.dumps(
        {
            "revocation_list": body,
            "signature": base64.urlsafe_b64encode(signature).decode().rstrip("="),
        }
    )

    assert _code_for(legacy, aid) == "signature_invalid"

    # Non-vacuity: the same body signed the CURRENT way does verify, so the
    # rejection above is about the signing input and not about the fixture
    # being malformed in some unrelated way.
    inner_input = json.dumps(body, separators=(",", ":"), sort_keys=True).encode()
    good_signature = private.sign(hashlib.sha256(inner_input).digest())
    current = json.dumps(
        {
            "revocation_list": body,
            "signature": base64.urlsafe_b64encode(good_signature).decode().rstrip("="),
        }
    )
    assert aitp.verify_revocation_list(current, aid) is None

File: bindings/aitp-node/tests/test_revocation.mjs
// Revocation-list signing — Node SDK.
//
// Mirrors bindings/aitp-py/tests/test_revocation.py: build a list with
// two entries, parse it back, and confirm the entries and expiry window
// survive the round-trip.
//
// Build first:  npm run build:debug
// Then run:     node --test tests/

import test from 'node:test';
import assert from 'node:assert/strict';
import { randomUUID } from 'node:crypto';

import { AitpAgent } from '../index.js';

function issuer() {
  const a = AitpAgent.generate();
  a.buildManifest({
    displayName: 'issuer',
    handshakeEndpoint: 'http://localhost:8100/aitp/handshake/',
    offeredCaps: ['demo.echo'],
  });
  return a;
}

test('signRevocationList round-trips two entries', () => {
  const iss = issuer();
  const jtiA = randomUUID();
  const jtiB = randomUUID();

  const envelopeJson = iss.signRevocationList(
    [
      { jti: jtiA, reason: 'compromised' },
      { jti: jtiB, revokedAt: 1_700_000_000 },
    ],
    600,
  );

  const env = JSON.parse(envelopeJson);
  const body = env.revocation_list;
  assert.equal(body.issuer, iss.aid);
  assert.equal(body.version, 'aitp/0.2');
  assert.equal(body.entries.length, 2);

  const jtis = new Set(body.entries.map((e) => e.jti));
  assert.deepEqual(jtis, new Set([jtiA, jtiB]));

  const byJti = Object.fromEntries(body.entries.map((e) => [e.jti, e]));
  assert.equal(byJti[jtiB].revoked_at, 1_700_000_000);
  assert.equal(byJti[jtiA].reason, 'compromised');

  // Expiry is published_at + 600.
  assert.equal(body.expires_at - body.published_at, 600);

  // Envelope carries a non-empty signature string.
  assert.equal(typeof env.signature, 'string');
  assert.ok(env.signature.length > 0);
});

test('signRevocationList rejects a bad UUID', () => {
  assert.throws(() => issuer().signRevocationList([{ jti: 'not-a-uuid' }]));
});

// ── Verification ─────────────────────────────────────────────────────────
//
// The verify half was missing until 0.6.0: both this binding and the Python
// one exposed signRevocationList and neither exposed verifyRevocationList,
// even though verify_revocation_snapshot is a Tier C conformance operation
// the Rust adapter implements. Three downstream repos then hand-rolled,
// skipped, or faked verification, and the 0.5.0 signing-input change crossed
// a whole release family with one accidental interlock in its way.

import { readFileSync } from 'node:fs';
import { generateKeyPairSync, createHash, sign as edSign } from 'node:crypto';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { verifyRevocationList, revocationSigningBytes } from '../index.js';

const REPO_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../..');

// RFC 8785 reduces to sorted-key compact JSON for these bodies: ASCII strings
// and integer numbers only, no floats, no escapes.
function canon(value) {
  if (Array.isArray(value)) return '[' + value.map(canon).join(',') + ']';
  if (value && typeof value === 'object') {
    return (
      '{' +
      Object.keys(value)
        .sort()
        .map((k) => JSON.stringify(k) + ':' + canon(value[k]))
        .join(',') +
      '}'
    );
  }
  return JSON.stringify(value);
}

function causeOf(fn) {
  try {
    fn();
  } catch (err) {
    // `err.code` is the contract. Deliberately NOT parsed out of
    // `err.message`: a caller string-matching an error message is pinning
    // program output as an expected value, which is the bug class this
    // binding exists to remove. If the cause channel ever regresses to a
    // message prefix, this returns 'GenericFailure' and every cause
    // assertion below fails loudly rather than silently passing.
    return err.code;
  }
  throw new Error('expected verification to fail, but it succeeded');
}

test('verifyRevocationList accepts a freshly signed snapshot', () => {
  const a = issuer();
  const env = a.signRevocationList([{ jti: randomUUID() }], 600);
  assert.equal(verifyRevocationList(env, a.aid), undefined);
});

test('verifyRevocationList verifies the committed spec vector as committed', () => {
  // Cross-implementation check: verified AS COMMITTED, never re-minted. A
  // suite where the same code signs and verifies passes under any
  // self-consistent convention — including a wrong one, which is how an
  // aitp-rs issuer and an aitp-verifier-py consumer disagreed on the wire
  // while both suites stayed green.
  const snap = JSON.parse(
    readFileSync(
      path.join(
        REPO_ROOT,
        'tests/schemas/known-answer/signed-examples/revocation/kat-keypair-001-snapshot.json',
      ),
      'utf8',
    ),
  );
  delete snap._kat_input; // a minting companion, never part of the wire object

  // Issuer comes from the keypair vector, NOT from the snapshot — deriving it
  // from the envelope would make the issuer-binding check tautological.
  const keypairs = JSON.parse(
    readFileSync(path.join(REPO_ROOT, 'tests/schemas/known-answer/keypairs.json'), 'utf8'),
  );
  const findAid = (node) => {
    if (Array.isArray(node)) return node.map(findAid).find(Boolean);
    if (node && typeof node === 'object') {
      if (node.id === 'kat-keypair-001' && node.aid) return node.aid;
      return Object.values(node).map(findAid).find(Boolean);
    }
    return undefined;
  };
  const aid = findAid(keypairs);
  assert.ok(aid, 'kat-keypair-001 not found in keypairs.json');

  assert.equal(verifyRevocationList(JSON.stringify(snap), aid, 1_711_900_100), undefined);
});

test('every failure cause is recoverable without parsing prose', () => {
  const a = issuer();
  const other = issuer();
  const env = a.signRevocationList([{ jti: randomUUID() }], 600);

  assert.equal(causeOf(() => verifyRevocationList(env, other.aid)), 'issuer_mismatch');

  const tampered = JSON.parse(env);
  tampered.revocation_list.entries = [];
  assert.equal(
    causeOf(() => verifyRevocationList(JSON.stringify(tampered), a.aid)),
    'signature_invalid',
  );

  const badVersion = JSON.parse(env);
  badVersion.revocation_list.version = 'aitp/9.9';
  assert.equal(
    causeOf(() => verifyRevocationList(JSON.stringify(badVersion), a.aid)),
    'version_unknown',
  );

  assert.equal(causeOf(() => verifyRevocationList(env, a.aid, 99_999_999_999)), 'expired');
  assert.equal(causeOf(() => verifyRevocationList('not json at all', a.aid)), 'malformed');
});

test('issuer_mismatch is distinct from malformed', () => {
  // The same cause until 0.6.0 (both CnfMalformed). An attacker serving their
  // own correctly-signed list and a corrupt fetch are different events; a
  // caller that cannot separate them cannot alert on the first without
  // drowning in the second.
  const a = issuer();
  const other = issuer();
  const env = a.signRevocationList([{ jti: randomUUID() }], 600);

  assert.equal(verifyRevocationList(env, a.aid), undefined);
  assert.equal(causeOf(() => verifyRevocationList(env, other.aid)), 'issuer_mismatch');
  assert.equal(causeOf(() => verifyRevocationList('{}', a.aid)), 'malformed');
});

test('a wrapped-form signature is rejected', () => {
  // The 0.5.0 break, from the verifier's side. Up to 0.4.x the signing input
  // was JCS({"revocation_list": body}) — the transport wrapper. The signer
  // here is node:crypto, deliberately independent of the code under test.
  const { privateKey, publicKey } = generateKeyPairSync('ed25519');
  const raw = publicKey.export({ type: 'spki', format: 'der' }).subarray(-32);
  const aid = 'aid:pubkey:' + raw.toString('base64url');

  const now = Math.floor(Date.now() / 1000);
  const body = {
    version: 'aitp/0.2',
    issuer: aid,
    published_at: now,
    expires_at: now + 600,
    entries: [{ jti: randomUUID(), revoked_at: now }],
  };

  const digest = (s) => createHash('sha256').update(Buffer.from(s)).digest();
  const wrappedSig = edSign(null, digest(canon({ revocation_list: body })), privateKey);
  const legacy = JSON.stringify({
    revocation_list: body,
    signature: wrappedSig.toString('base64url'),
  });
  assert.equal(causeOf(() => verifyRevocationList(legacy, aid)), 'signature_invalid');

  // Non-vacuity: the same body signed the CURRENT way DOES verify, so the
  // rejection above is about the signing input, not a malformed fixture.
  const innerSig = edSign(null, digest(canon(body)), privateKey);
  const current = JSON.stringify({
    revocation_list: body,
    signature: innerSig.toString('base64url'),
  });
  assert.equal(verifyRevocationList(current, aid), undefined);
});

test('revocationSigningBytes are the inner body, not the wrapper', () => {
  const a = issuer();
  const env = a.signRevocationList([{ jti: randomUUID() }], 600);
  const bytes = revocationSigningBytes(env);
  assert.ok(bytes.length > 0);
  assert.equal(bytes[0], '{'.charCodeAt(0));
  assert.ok(
    !bytes.subarray(0, 32).toString().includes('revocation_list'),
    'signing input starts with the transport wrapper — that is the pre-0.5.0 shape',
  );
  // Derived from the envelope, never pasted from output.
  assert.equal(bytes.toString(), canon(JSON.parse(env).revocation_list));
});
